src/librustc_mir/interpret/eval_context.rs

```diff
@@ -383,40 +383,29 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
     ) -> EvalResult<'tcx> {
         ::log_settings::settings().indentation += 1;
 
-        /// Return the set of locals that have a storage annotation anywhere
-        fn collect_storage_annotations<'mir, 'tcx>(mir: &'mir mir::Mir<'tcx>) -> HashSet<mir::Local> {
-            use rustc::mir::StatementKind::*;
-
-            let mut set = HashSet::new();
-            for block in mir.basic_blocks() {
-                for stmt in block.statements.iter() {
-                    match stmt.kind {
-                        StorageLive(local) |
-                        StorageDead(local) => {
-                            set.insert(local);
-                        }
-                        _ => {}
-                    }
-                }
-            }
-            set
-        }
-
         // Subtract 1 because `local_decls` includes the ReturnMemoryPointer, but we don't store a local
         // `Value` for that.
         let num_locals = mir.local_decls.len() - 1;
 
-        let locals = {
-            let annotated_locals = collect_storage_annotations(mir);
-            let mut locals = vec![None; num_locals];
-            for i in 0..num_locals {
-                let local = mir::Local::new(i + 1);
-                if !annotated_locals.contains(&local) {
-                    locals[i] = Some(Value::ByVal(PrimVal::Undef));
+        let mut locals = vec![Some(Value::ByVal(PrimVal::Undef)); num_locals];
```

Comment:
Please use `IndexVec` here.

Reply:
The reason we didn't use that is theat we don't actually use the return pointer local at all.

Reply:
Fixed, the waste of one space massively outweighs the many `.index() - 1` calls we had.

Reply:
Are you sure? Now you have to allocate something on the heap even if it's never used.

Reply:
For truly trivial constants (`const FOO: usize = 42;`) that is true, but even `const FOO: usize = 40 + 2;` will create that heap alloc.

I added a fast path for trivial constants

Reply:
:+1: